Upgrade Typescript to v5

sisou opened the pull request:
v4 -> v5 was no changes actually, see the commit history.

~This is based on the `usdc` branch, so can only be merged after #465 is merged.~

sisou committed:
Update Typescript to latest v3

sisou committed:
Update to Typescript v5

sisou committed:
Remove TS compatibility fix

sisou committed:
Update to Typescript v4

onmax commented on src/components/RecoveryWords.js:
```diff
@@ -157,9 +165,10 @@ class RecoveryWords extends Nimiq.Observable {
-                this.$fields[index].fillValueFrom(paste);
+            const field = this.$fields[index];
+            field.focus();
+            if (paste && 'fillValueFrom' in field) {
```
```suggestion
            if (paste && 'fillValueFrom' in field && typeof field.fillValueFrom === 'function') {
```

not sure if this is relevant

sisou replied:
I don't think this is necessary.

onmax commented on src/components/RecoveryWords.js:
```diff
@@ -123,18 +130,19 @@ class RecoveryWords extends Nimiq.Observable {
-            if (e.message !== 'Invalid checksum') {
+            const err = /** @type {Error} */ (e);
+            if (err.message !== 'Invalid checksum') {
                 console.error(e); // eslint-disable-line no-console
```
```suggestion
                console.error(err); // eslint-disable-line no-console
```

_Super picky comment_

I wouldn't use `e` after defining `err` with a type

onmax commented on tsconfig.json:
```diff
@@ -1,8 +1,8 @@
     /* Basic Options */
-    "target": "ES2015",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
-    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "target": "ES2020",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
```
```suggestion
    "target": "ES2020",                       /* Specify ECMAScript target version. */
```

onmax replied:
People can learn what values are valid from the docs so we don't have these comments out of sync ?

onmax commented on tsconfig.json:
```diff
@@ -1,8 +1,8 @@
-    "target": "ES2015",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
-    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "target": "ES2020",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
+    "module": "CommonJS",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
```
```suggestion
    "module": "CommonJS",                     /* Specify module code generation. */
```

onmax reviewed (approved):
LGTM

danimoh committed:
LoginFile: avoid swallowing errors in canvas.toBlob

danimoh committed:
Typechecks instead of type assertions for caught errors

danimoh committed:
Typecheck service worker separately to avoid DOM and Webworker lib type clashs

Typecheck service worker separately with ts WebWorker lib, and the rest of the Keyguard
with ts DOM lib. Previously, both libs were active in all files (even though webworker
was referenced only in the ServiceWorker via a triple slash directive, this enables it
globally), which by itself is inappropriate, but also causes type clashs between the
DOM and the WebWorker types, which previously were silenced by skipping lib checks.

Now, with this change, and typescript having been updated to a recent version, checking
libs has been enabled again.

danimoh committed:
Remove @ts-ignores resolved with updated ts; change others to @ts-expect-error

danimoh committed:
tsconfig: bump es target version

danimoh committed:
IqonHash: remove String.padEnd polyfill

Browser support for padEnd is now sufficiently good.

danimoh committed:
Consistent typing of $els in components

Allow them to be passed as null or undefined everywhere.

danimoh committed:
client: update typescript version

danimoh reviewed (approved):
Cool, thanks for this.
I added my improvements, feel free to have a look.
The change of static properties of classes to actual js `static`s requires an eslint update, so this will be in a separate PR.

@onmax please also review my changes, if you want.

danimoh commented:
The force-push is for rebase onto master, with resolved conflicts in the `FiatApi`.
However, some additional type errors seem to have surfaced on the latest changes from master.
Taking a look now.

sisou commented on src/lib/LoginFile.js:
```diff
@@ -65,9 +65,12 @@ class LoginFile {
+                if (!blob) {
+                    reject(new Error('Cannot generate URL'));
+                    return;
+                }
```
TIL throwing in the `new Promise()` callback doesn't reject the promise. The more you know...

sisou commented on tsconfig.json:
```diff
@@ -4,11 +4,6 @@
-                                              /* lib.webworker.d.ts referenced in ServiceWorker.js clash with types */
-                                              /* in lib.dom.d.ts, and also because our typescript version is too old */
-                                              /* to handle the types of ethers. */
-                                              /* TODO once typescript has been updated, this can be removed. */
```
Well spotted! :+1:

sisou commented on src/common.js:
```diff
@@ -88,8 +88,7 @@ async function runKeyguard(RequestApiClass, opts) { // eslint-disable-line no-un
     document.body.addEventListener('click', event => {
-        // @ts-ignore (Property 'matches' does not exist on type 'EventTarget'.)
-        if (!event.target || !event.target.matches('a.page-header-back-button')) return;
+        if (!(event.target instanceof HTMLElement && event.target.matches('a.page-header-back-button'))) return;
```
I liked the previous double-negative OR syntax better.

danimoh replied:
Me theoretically too, I only changed it to make it fit into a single line.

sisou replied:
Changed it back to double negative, that version is only one character longer and still fits into one line.

danimoh replied:
Oh, right :+1:
Sometimes in my IDE lines look longer when it shows names of the passed parameters.

sisou commented on src/components/PasswordBox.js:
```diff
@@ -41,7 +41,7 @@ class PasswordBox extends Nimiq.Observable {
         this.$el.classList.toggle('hide-input', this.options.hideInput);
 
-        this._passwordInput = new PasswordInput(this.$el.querySelector('[password-input]'));
+        this._passwordInput = new PasswordInput(/** @type {HTMLElement} */(this.$el.querySelector('[password-input]')));
```
Why is this change necessary? I thought with your changes `null | undefined` are also allowed?

danimoh replied:
I think that it worked previously, is actually a typescript bug. `querySelector` returns `Element | null`, `PasswordInput` previously accepted `?HTMLElement` which is equivalent to `HTMLElement | null`, which should refuse a passed value of `Element | null` but didn't. Now it's `{?HTMLElement} [$el]`, which is equivalent to `HTMLElement | null | undefined`, which somehow trigger typscript to actually complain (i.e. it's now actually more restrictive), which is why a type assertion is now required. This is the way it should be imo.

sisou commented on src/components/PasswordSetterBox.js:
```diff
@@ -27,7 +27,7 @@ class PasswordSetterBox extends Nimiq.Observable {
 
         this._passwordInput = new PasswordInput(
-            this.$el.querySelector('[password-input]'),
+            /** @type {HTMLElement} */ (this.$el.querySelector('[password-input]')),
```
Same here.

danimoh replied:
https://github.com/nimiq/keyguard/pull/467#discussion_r1746968333

sisou commented on src/components/PasswordSetterBox.js:
```diff
@@ -52,7 +52,7 @@ class PasswordSetterBox extends Nimiq.Observable {
 
     /**
-     * @param {?HTMLFormElement} $el
+     * @param {?HTMLFormElement | undefined} $el
```
What's the difference between `| undefined` in the type and putting `$el` in square brackets?

danimoh replied:
`[$el]` is an optional parameter which can be omitted. A parameter with `| undefined` can be set to `undefined` but not omitted. An optional parameter can not precede a non-optional parameter, which is why here it needs to be declared as `| undefined`.

In the old ts version, there was also another difference, which I now didn't observe anymore with the newer version: https://github.com/nimiq/keyguard/pull/461#discussion_r1045729111

danimoh committed:
FiatApi: assign enum values and types based on the provider

This was already the case before updating typescript to the latest v3 version in 35859ede,
which temporarily removed that handling as is was not compatible with newer typescript
versions. This commit reimplements this behavior in a different fashion.

sisou committed:
Wrapped negative AND to double negative OR

For readability and easier understanding.
